app/blog/tanstack-router-typescript-performance.md

```diff
@@ -0,0 +1,214 @@
+
+Despite all these past efforts (which certainly helped), we had to address the elephant in the room. The fundamental problem to solve for a great editor experience in TanStack Router was not necessarily related to overall typescript check time. The problem we have been working to resolve is the accumulated route tree being a type checking bottleneck for the language service. For those familiar with tracing TypeScript, a trace for a large TanStack Router applications could look something similar to the following.
+
+<img src="/blog-tracing-slow.png">
+
+For those who don't know, you can generate a trace from typescript with the following
+
+```
+tsc --generatetrace trace
+```
+
+This example has 400 route definitions all with `validateSearch` using `zod` and query integration through `context` and `loader`'s - it is an extreme example. The large wall at the beginning of the trace is what TypeScript was type checking when it first hits `Link`. The language service type checks a file or region of a file from the beginning but only that region or file. This meant the language service also had to do this work whenever you interact with `Link`. It turns out this wall is TypeScript inferring all the types in all the route definitions of the whole route tree. As mentioned, route definitions can in turn contain complex types from external validation libraries which are also inferred. This is quite clearly going to slow down the editing experience.
+
+## Breaking down work for the language service
+
+Ideally the language service should only need to infer from a route definition a `Link` is navigating to and not the whole route tree. This way the language service only needs to do work necessary to infer types from a single route definition a `Link` navigates to. Unfortunately code based route trees rely on inference to build the route tree which triggers the wall shown above. However with file based, the route tree is generated when a file is created or modified. Previously route trees were created like the following, even for file based.
+
+```tsx
+export const routeTree = rootRoute.addChildren({
+  IndexRoute,
+  LayoutRoute: LayoutRoute.addChildren({
+    LayoutLayout2Route: LayoutLayout2Route.addChildren({
+      LayoutLayout2LayoutARoute,
+      LayoutLayout2LayoutBRoute,
+    }),
+  }),
+  PostsRoute: PostsRoute.addChildren({ PostsPostIdRoute, PostsIndexRoute }),
+})
+```
+
+Generating the route tree came as a consequence of reducing tedious configuration of a route tree but keeping inference where it matters. This is where the first important change is introduced leading to better editor performance. Instead of inferring the route tree, we can take advantage of this generation step to _declare the route tree_.
```

Comment:
```suggestion
This generation process was initially introduced to reduce the tedious complexity involved in configuring your route tree, whilst still keeping inference where it mattered. This is where the first important change was introduced. Instead of inferring the route tree in a single shot, we can take advantage of this generation step to _declare the route tree_.
```